pkg/machine/compression/copy.go

```diff
@@ -0,0 +1,117 @@
+package compression
+
+import (
+	"bytes"
+	"io"
+	"os"
+)
+
+// TODO vendor this in ... pkg/os directory is small and code should be negligible
+/*
+	NOTE:   copy.go and copy.test were lifted from github.com/crc-org/crc because
+			i was having trouble getting go to vendor it properly. all credit to them
+*/
+
+func copyFile(src, dst string, sparse bool) error {
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+
+	defer in.Close()
+
+	out, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+
+	defer out.Close()
+
+	if sparse {
+		if _, err = CopySparse(out, in); err != nil {
+			return err
+		}
+	} else {
+		if _, err = io.Copy(out, in); err != nil {
+			return err
+		}
+	}
+
+	fi, err := os.Stat(src)
+	if err != nil {
+		return err
+	}
+
+	if err = os.Chmod(dst, fi.Mode()); err != nil {
+		return err
+	}
+
+	return out.Close()
+}
+
+func CopyFile(src, dst string) error {
+	return copyFile(src, dst, false)
+}
+
+func CopyFileSparse(src, dst string) error {
+	return copyFile(src, dst, true)
+}
+
+func CopySparse(dst io.WriteSeeker, src io.Reader) (int64, error) {
+	copyBuf := make([]byte, copyChunkSize)
+	sparseWriter := newSparseWriter(dst)
+
+	bytesWritten, err := io.CopyBuffer(sparseWriter, src, copyBuf)
+	if err != nil {
+		return bytesWritten, err
+	}
+	err = sparseWriter.Close()
+	return bytesWritten, err
+}
+
+type sparseWriter struct {
+	writer          io.WriteSeeker
+	lastChunkSparse bool
+}
+
+func newSparseWriter(writer io.WriteSeeker) *sparseWriter {
+	return &sparseWriter{writer: writer}
+}
+
+const copyChunkSize = 4096
+
+var emptyChunk = make([]byte, copyChunkSize)
+
+func isEmptyChunk(p []byte) bool {
+	// HasPrefix instead of bytes.Equal in order to handle the last chunk
+	// of the file, which may be shorter than len(emptyChunk), and would
+	// fail bytes.Equal()
+	return bytes.HasPrefix(emptyChunk, p)
+}
+
+func (w *sparseWriter) Write(p []byte) (n int, err error) {
+	if isEmptyChunk(p) {
+		offset, err := w.writer.Seek(int64(len(p)), io.SeekCurrent)
+		if err != nil {
+			w.lastChunkSparse = false
+			return 0, err
+		}
+		_ = offset
+		w.lastChunkSparse = true
+		return len(p), nil
```

Comment:
should we handle each byte separately so we can detect sequences of zeros that are within a single frame?

For example, if the Write method is called with a buffer consisting of a 2MB sequence of zeros and the last byte is not equal to zero, wouldn't the current code fail to detect it?

Reply:
the code appears to work and has been used in crc for some time; however, we fully intend to remove the copied file and actually vendor the code then.  So, if we have suggestions, we could take them to crc-org/crc/pkg/os ?

Reply:
crc uses this only through `CopySparse` which uses `CopyBuffer` to ensure `sparseWriter.Write()` is called with 4096 bytes chunks.
2MB sequence of zeros and non-zero last byte will indeed not be detected appropriately as being sparse.
This code could definitely be improved by iterating over the write buffer without relying on the calling code having 4k granularity. 64, 512 or 4k granularity for sparse detection might be less heavyweight than doing it byte by byte from a performance point of view.